Samples/ResourceManagement/cs-winforms-unpackaged/Program.cs

```diff
@@ -1,54 +1,60 @@
-            // The resource manager is a superset of the resource loader in terms of functionality.
-            Application.Run(new AppForm(resourceManager, resourceLoader, overrideResourceContext));
-
-            // Uninitialize dynamic dependencies.
-            MddBootstrap.Shutdown();
-        }
-    }
-}
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Windows.Forms;
+using Microsoft.ApplicationModel.Resources;
+using Microsoft.Windows.ApplicationModel;
+
+namespace winforms_unpackaged_app
+{
+    static class Program
+    {
+        /// <summary>
+        ///  The main entry point for the application.
+        /// </summary>
+        [STAThread]
+        static void Main()
+        {
+            Application.SetHighDpiMode(HighDpiMode.SystemAware);
+            Application.EnableVisualStyles();
+            Application.SetCompatibleTextRenderingDefault(false);
+
+            // Initialize dynamic dependencies so we can consume the Project Reunion APIs in the Project
+            // Reunion framework package from this unpackaged app.
+            // Take a dependency on Project Reunion v0.8 preview.
+            int hr = MddBootstrap.Initialize(8, "preview");
+            if (hr < 0)
+            {
+                MessageBox.Show("Fail to initialize Dynamic Dependencies with error " + hr, "Error!");
+                return;
+            }
+            
+            var resourceManager = new ResourceManager("winforms_unpackaged_app.pri");
+            // Fall back to other resource loaders if the resource is not found in MRT, in this case .net.
+            // This enables carrying forward existing assets without the need to convert them.
+            resourceManager.ResourceNotFound += (sender, args) =>
+            {
+                var candidate = new ResourceCandidate(
+                    ResourceCandidateKind.String,
+                    LegacyResources.ResourceManager.GetString(args.Name, LegacyResources.Culture));
+
+                args.SetResolvedCandidate(candidate);
+            };
+
+            var resourceLoader = new ResourceLoader();
+
+            // Create a custom resource context. Set the language to German.
+            var overrideResourceContext = resourceManager.CreateResourceContext();
+            overrideResourceContext.QualifierValues["Language"] = "de-DE";
+
+            // The resource manager and the resource loader do disk I/O and should be centrally cached.
+            // Since both end up loading the same file, ideally an app should only use one.
+            // The resource manager is a superset of the resource loader in terms of functionality.
+            Application.Run(new AppForm(resourceManager, resourceLoader, overrideResourceContext));
+
+            // Uninitialize dynamic dependencies.
+            MddBootstrap.Shutdown();

```

Comment:
@DrusTheAxe it occurred to me that requiring an explicit shutdown call is prone to leak in exception-based languages. Do you have recommendation for handling this? Does this object unload on scope exit, or does it have a RAII wrapper?

Reply:
I think the `MddBootstrap` object is part of the sample program itself: https://github.com/microsoft/WindowsAppSDK-Samples/blob/e6d6bd953a6498b9efd510de80591edd862ae8f8/Samples/ResourceManagement/cs-winforms-unpackaged/MddBootstrap.cs#L49

Maybe/Perhaps that class could be updated to be more of a RAII object and call `Shutdown` when destructed?

Reply:
Good catch! Yes, the class should be RAII-ified. In fact we should probably create the canonical C# wrapper for this, and a canonical C++ RAII wrapper for this. Otherwise every app will create its own.

Reply:
MddBootstrap.cs was a quick helper thrown together for initial samples. The Bootstrap C# API is on its way...

[Spec updates for GenerationId and Bootstrap C# API #1284](https://github.com/microsoft/WindowsAppSDK/pull/1284)

[Add Bootstrap C# API. Spec Bootstrap C# and get GenerationId (Flat-C/WinRT) APIs #1274](https://github.com/microsoft/WindowsAppSDK/pull/1274)

Once complete samples can (should) use the API and delete their copy of MddBootstrap.cs

Reply:
> canonical C++ RAII wrapper

A C++ RAII wrapper isn't planned as the value seemed low (and in some ways, negative). Given WIL, `RETURN_IF_FAILED(MddBootstrapInitialize(...))` or `THROW_IF_FAILED(MddBootstrapInitialize(...))` didn't see a problem.

As for RAII, that is questionable and potentially dangerous. `MddBootstrapShutdown()` reverts what `MddBootstrapInitialize ()` did, which is mainly memory fiddling. As this is expected to be done once at the process level (e.g. in main/WinMain) then unfiddling memory right before the process ends isn't important. The key thing MddBootstrapInitialize does is to create a running process with the identity of a DDLM package (to ensure Windows doesn't service the WinAppSDK Framework while you're using it). That process is kept running until `MddBootstrapShutdown`, which ends it.

Today that process is a PackagedCOM OOP server, so shutdown is `p->Release()` to explicitly stop using it (and if this was the only process with a reference, to end the process). If you don't call it and just ExitProcess() Windows will cleanup your process' references to OOP COM servers. Calling `MddBootstrapShutdown` is slightly more efficient than ending your process without calling it, but same net result.

So, no harm no foul if you don't call `MddBootstrapShudown()` when your process is ending.

Given a RAII wrapper `MddBootstrapShutdown` would be called by the object's destructor when the object goes out of scope. Which sounds good...unless you have anything using WinAppSDK executing after that. That could be harmless to problematic to fatal, and easily can lead to subtle and cryptic problems. So, maybe harmless, maybe not so.

Hence the reason the Bootstrap API was provided as a C API for C++ developers.

Sure it's possible we could provide a C++ RAII wrapper. What benefit do you think that would provide? And does that outweigh the potential downsides?

NOTE: The Bootstrap C# API exists for multiple reasons (some of which don't apply to C++). See the links above for more details.

Reply:
> requiring an explicit shutdown call is prone to leak

Usually...but not here.

Failure to call MddBootstrapShutdown() at process termination means Windows' process cleanup will drop the reference to the DDLM OOP COM server. No leak, just a slight difference in timing and codepath.

It also means we won't unfiddle the memory fiddled with by `MddBootstrapInitialize` (altered PATH environment variable, etc). But if the process is terminating then that's harmless.

And the Bootstrap API is designed to be called once before/after all WinAppSDK usage in a process. In practice that means at the very start and end of `main()`/`WinMain()`.

I'm a huge RAII supporter. Call this the exception providing the rule

(though I'm open to feedback if folks think a C++ RAII wrapper is helpful)

Reply:
Given Howard's argument, I'll leave the class as it is. Further discussion and a decision may be pending, but that can be done separately and any action needed can be taken separately.